dev/ProjectReunion_BootstrapDLL/MddBootstrap.cpp

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+
+#include "MddBootstrap.h"
+
+#include "IDynamicDependencyLifetimeManager.h"
+
+HRESULT GetFrameworkPackageInfoForPackage(PCWSTR packageFullName, const PACKAGE_INFO*& frameworkPackageInfo, wil::unique_cotaskmem_ptr<BYTE[]>& packageInfoBuffer);
+HRESULT AddFrameworkToPath(PCWSTR  path);
+
+IDynamicDependencyLifetimeManager* g_lifetimeManager = nullptr;
+DLL_DIRECTORY_COOKIE g_dllDirectoryCookie = 0;
+wil::unique_cotaskmem_string g_frameworkPath;
+
+STDAPI MddBootstrapInitialize(
+    const CLSID& appDynamicDependencyLifetimeManager)
+{
+    wil::com_ptr_nothrow<IDynamicDependencyLifetimeManager> lifetimeManager;
+    RETURN_IF_FAILED(CoCreateInstance(appDynamicDependencyLifetimeManager, nullptr, CLSCTX_LOCAL_SERVER, __uuidof(IDynamicDependencyLifetimeManager), reinterpret_cast<void**>(lifetimeManager.addressof())));
+
+    RETURN_IF_FAILED(lifetimeManager->Initialize());
+
+    wil::unique_cotaskmem_string packageFullName;
+    RETURN_IF_FAILED(lifetimeManager->GetPackageFullName(&packageFullName));
+
+    wil::unique_cotaskmem_ptr<BYTE[]> packageInfoBuffer;
+    const PACKAGE_INFO* frameworkPackageInfo = nullptr;
+    RETURN_IF_FAILED(GetFrameworkPackageInfoForPackage(packageFullName.get(), frameworkPackageInfo, packageInfoBuffer));
+
+    RETURN_IF_FAILED(AddFrameworkToPath(frameworkPackageInfo->path));
+
+    g_lifetimeManager = lifetimeManager.detach();
+    return S_OK;
+}
+
+STDAPI_(void) MddBootstrapShutdown()
+{
+    if (!g_lifetimeManager)
+    {
+        (void)LOG_IF_FAILED(g_lifetimeManager->Shutdown());
+        g_lifetimeManager->Release();
+        g_lifetimeManager = nullptr;
+    }
+
+    if (g_dllDirectoryCookie != 0)
+    {
+        RemoveDllDirectory(g_dllDirectoryCookie);
+        g_dllDirectoryCookie = 0;
+    }
+
+    if (g_frameworkPath)
+    {
+        wil::unique_cotaskmem_string path;
+        const auto hr = wil::TryGetEnvironmentVariableW(L"PATH", path);
+        if (SUCCEEDED(hr) && path)
+        {
+            const auto frameworkPathLength = wcslen(g_frameworkPath.get());
+            PCWSTR pathWithoutFrameworkPath = g_frameworkPath.get() + frameworkPathLength;
+            if (*pathWithoutFrameworkPath == L';')
+            {
+                ++pathWithoutFrameworkPath;
+            }
+            (void)LOG_IF_WIN32_BOOL_FALSE(SetEnvironmentVariableW(L"PATH", pathWithoutFrameworkPath));
```

Comment:
Doesn't this assume that only the MDD code is modifying %path% so it's trimming only the start/end? Maybe this should use std::wstring to manipulate the path string, and delete the first instance of the path it found.

Alternatively, I just noticed, does this assume a single target framework package? What if we wanted Reunion and Microsoft.AI to be peer packages? Could this code be reused?

Reply:
`Alternatively` - The intent is to alter path just long enough to LoadLibrary(Microsoft.ProjectReunion.dll) from the PRframework package. That'll initialize the package graph, adding Reunion to it. There's a bit of a chicken-egg problem I haven't entirely settled out which way I want to handle it. I need to get more details working with tests before finalizing. For now this is enough to get a bootstrap UT working, so I can dive into the deeper DynDep waters. I expect to come back here once I get MddAddPackageDependnecy's UT running